### PR TITLE
rgw: lua, add support for lua packages to the lua manager

### PR DESCRIPTION
**Pull request by kalebskeithley**
Refactor lua manager, adding new {add,remove}_package(), and
list_packages() methods; and changing get(), put(), del() methods
to {get,put,del}_script().  Renaming class from *LuaScriptManager
to *LuaManager to be consistent with dual *script and *package
functionality.

Note: I considered keeping the existing *LuaScriptManager classes
as is and adding *LuaPackageManager classes, but that seemed/seems
like overkill to me.

Signed-off-by: Kaleb S. KEITHLEY <kkeithle@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by yuvalif — APPROVED** [review]

**Review by dang — APPROVED** [review]

File: src/rgw/rgw_sal_store.h [hunk 1]
```diff
@@ -404,9 +404,22 @@ class StoreZone : public Zone {
...
+  /** Put a script named with the given key to the backing store */
+  virtual int put_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) { return -ENOENT; }
+  /** Delete a script named with the given key from the backing store */
+  virtual int del_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) { return -ENOENT; }
+  /** Add a lua package */
+  virtual int add_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name) { return -ENOENT; }
+  /** Remove a lua package */
+  virtual int remove_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name) { return -ENOENT; }
+  /** List lua packages */
+  virtual int list_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& packages) { return -ENOENT; }
```
Comment (cbodley) [inline src/rgw/rgw_sal_store.h:422]: can we please leave these as pure virtual (i.e. not override them here at all), so every derived store remembers to implement them?

File: src/rgw/rgw_sal_filter.cc [hunk 2]
```diff
@@ -1303,22 +1303,40 @@ int FilterWriter::complete(size_t accounted_size, const std::string& etag,
...
 				const std::string& key)
 {
-  return next->del(dpp, y, key);
+  return next->del_script(dpp, y, key);
+}
+
+int FilterLuaManager::add_package(const DoutPrefixProvider* dpp, optional_yield y,
+                                 const std::string& package_name)
+{
+  return -ENOENT;
```
Comment (cbodley) [inline src/rgw/rgw_sal_filter.cc:1327]: filters are supposed to forward all calls to the `next` store. this should `return next->add_package(...)`. same for remove_ and list_packages()

**Comment by cbodley** [comment]
in the future, please just force-push your branch instead of closing/opening a new PR - that way we can keep all of the review history together

File: src/rgw/rgw_sal_store.h (hunk 1 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal_store.h]: done

File: src/rgw/rgw_sal_filter.cc (hunk 2 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal_filter.cc]: done

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
i would really prefer that we not merge things without actual teuthology results

**Comment by ljflores** [comment]
@kalebskeithley @cbodley can you take a look at https://tracker.ceph.com/issues/57079? It's a dependency issue that looks related to this PR, which only occurs on certain builds. The commonality is that they all are built on adami06, which has the very new Ubuntu 22.04 (Jammy Jellyfish) OS.

**Comment by ljflores** [comment]
> @kalebskeithley @cbodley can you take a look at https://tracker.ceph.com/issues/57079? It's a dependency issue that looks related to this PR, which only occurs on a handful of make check builds. The commonality is that they all are built on adami06, which has the very new Ubuntu 22.04 (Jammy Jellyfish) OS.

NVM, there is already a fix! #47517

**Pull request merged**